File: 108.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.empty()) return nullptr; 
        int mid = nums.size()/2; 
        TreeNode* head = new TreeNode(nums[mid]); 
        
        vector <int> left (nums.begin(), nums.begin()+mid); 
        vector<int>right(nums.begin()+mid+1, nums.end()); 
        head->right = sortedArrayToBST(right); 
        head->left = sortedArrayToBST(left); 
        
        return head; 
        
    }
    
};
//Even better to use iterators 

/*
class Solution {
public:
    using iter = vector<int>::const_iterator;
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.empty()) return nullptr;
        
        return buildBST(nums.begin(), nums.end());
    }
    
    TreeNode* buildBST(iter b, iter e) {
        if (b >= e) return nullptr;
        
        iter m = b + (e - b) / 2;
        
        TreeNode* node = new TreeNode(*m);
        node->left = buildBST(b, m);
        node->right = buildBST(m + 1, e);
        
        return node;
    }
};
*/ 

File: 108.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        return self.calculate(nums, 0, len(nums))
    
    def calculate(self, nums, start, end):
        middle_pos = (end-start)//2+start

        if end < start or end < 0 or start == len(nums):
            return None 
        elif end == start:
            return TreeNode(nums[middle_pos])
        
        left_node = self.calculate(nums, start, middle_pos-1)
        right_node = self.calculate(nums, middle_pos+1, end)
            
        middle_node = TreeNode(nums[middle_pos], left_node, right_node)
        
        return middle_node
